src/hyperlight_host/src/hypervisor/mod.rs

```diff
@@ -159,26 +167,264 @@ pub(crate) trait Hypervisor: Debug + Send {
+    #[allow(clippy::too_many_arguments)]
     fn run(
         &mut self,
-        #[cfg(feature = "trace_guest")] tc: &mut crate::sandbox::trace::TraceContext,
-    ) -> Result<HyperlightExit>;
+        _entrypoint: u64,
+        interrupt_handle: Arc<dyn InterruptHandleImpl>,
+        sandbox_regions: &[MemoryRegion],
+        mmap_regions: &[MemoryRegion],
+        mem_mgr: &mut SandboxMemoryManager<HostSharedMemory>,
+        host_funcs: Arc<Mutex<FunctionRegistry>>,
+        #[cfg(gdb)] dbg_mem_access_fn: Arc<Mutex<SandboxMemoryManager<HostSharedMemory>>>,
+        #[cfg(crashdump)] rt_cfg: &SandboxRuntimeConfig,
+    ) -> Result<()> {
+        // ===== KILL() TIMING POINT 1: Between guest function calls =====
+        // Clear any stale cancellation from a previous guest function call or if kill() was called too early.
+        // This ensures that kill() called BETWEEN different guest function calls doesn't affect the next call.
+        //
+        // If kill() was called and ran to completion BEFORE this line executes:
+        //    - kill() has NO effect on this guest function call because CANCEL_BIT is cleared here.
+        //    - NOTE: stale signals can still be delivered, but they will be ignored.
+        interrupt_handle.clear_cancel();
+
+        // Keeps the trace context and open spans
+        #[cfg(feature = "trace_guest")]
+        let mut tc = crate::sandbox::trace::TraceContext::new();
+
+        let result = loop {
+            // ===== KILL() TIMING POINT 2: Before set_tid() =====
+            // If kill() is called and ran to completion BEFORE this line executes:
+            //    - CANCEL_BIT will be set and we will return an early VmExit::Cancelled()
+            interrupt_handle.set_tid();
+            interrupt_handle.set_running();
+
+            let exit_reason =
+                if interrupt_handle.is_cancelled() || interrupt_handle.is_debug_interrupted() {
+                    Ok(VmExit::Cancelled())
+                } else {
+                    #[cfg(feature = "trace_guest")]
+                    tc.setup_guest_trace(tracing_opentelemetry::OpenTelemetrySpanExt::context(
+                        &tracing::Span::current(),
+                    ));
 
-    /// Get InterruptHandle to underlying VM (returns internal trait)
-    fn interrupt_handle(&self) -> Arc<dyn InterruptHandleInternal>;
+                    // ===== KILL() TIMING POINT 3: Before calling run_vcpu() =====
+                    // If kill() is called and ran to completion BEFORE this line executes:
+                    //    - CANCEL_BIT will be set, but it's too late to prevent entering the guest this iteration
+                    //    - Signals will interrupt the guest (RUNNING_BIT=true), causing VmExit::Cancelled()
+                    //    - If the guest completes before any signals arrive, kill() may have no effect
+                    //      - If there are more iterations to do (IO/host func, etc.), the next iteration will be cancelled
```

Comment:
What happens if there are not more iterations to do , is it possible to lie about completion or cancel the next use of this Sandbox?